### ISSUE TITLE
409 Error when RClone Sync --delete-before to Azure Blob Archive-Tier

### ISSUE THREAD

**Issue by GoingOffRoading** [rclone/rclone#4819]
#### What is the problem you are having with rclone?

I have files in Azure Blob storage that are stored in Archive-Tier.  Azure Blob Archive-Tier files can be stored or deleted, have meta-data, and they CAN-NOT be read or modified.  My local versions of these files have been updated and I now need to sync my local files to Azure.  To do so, I essentially need to delete the Azure Blob version (in Archive-Tier, again: can not be read/modified outside of meta-data).

I'm using the RClone GUI but am setting the --delete-before flag on the sync.

When the sync occurs, I get a "RESPONSE Stats: 409 This operation is not permitted on an archived blob" error.  The error code is many rows but those code do not provide any additional diagnostic information, just Azure, API, local machine, and etc basic information.

Note, similar issue here: [https://forum.rclone.org/t/azure-storage-blob-overwrite-on-archive/9668](https://forum.rclone.org/t/azure-storage-blob-overwrite-on-archive/9668)

#### What is your rclone version (output from `rclone version`)

```
# rclone version
rclone v1.53.0
- os/arch: linux/amd64
- go version: go1.15
```

Using the GUI option

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Running in Docker/Kubernetes, linux/amd64

####  Which cloud storage system are you using? (e.g. Google Drive)

Azure Blob Storage

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone sync --delete-before --size-only --verbose --no-update-modtime --transfers 4 --checkers 8 --contimeout 60s --timeout 300s --retries 3 --low-level-retries 10 --delete-before  --stats 1s --stats-file-name-length 0 --fast-list /directory:remote`

Please forgive small formatting errors... This was manually copied from a VNC window

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

Deleted/modified the directory, remote name, and the GUIDs for privacy reasons

```
2020/11/30 01:21:25 ERROR : **MY-FILE**: Failed to copy: multipart upload failed to upload part: -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=BlobArchived) =====
Description=This operation is not permitted on an archived blob.
RequestId:c97536a8-e01e-000b-50b7-c6f95f0000
Time:2020-11-30T01:21:25.4901576Z, Details:
   Code: BlobArchived
   PUT https://kubeomv.blob.core.windows.net/**MyRemote**?blockid=AQAAAAAAAAA%3D&comp=block&timeout=31536001
   Authorization: REDACTED
   Content-Length: [419434]
   Content-Md5: [km02rBUA9Ac7ArHhzRuw==]
   User-Agent: [rclone/v1.53.0]
   X-Ms-Client-Request-Id: [9409ff8f-f83f-4fae-53cc-556e67c74852]
   X-Ms-Date: [Mon, 30 Nov 2020 01:21:25 GMT]
   X-Ms-Version: [2019-02-02]
   --------------------------------------------------------------------------------
   RESPONSE Status: 409 This operation is not permitted on an archived blob.
   Content-Length: [233]
   Content-Type: [application/xml]
   Date: [Mon, 30 Nov 2020 01:21:24 GMT]
   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]
   X-Ms-Client-Request-Id: [9409ff8f-f83f-4fae-53cc-556e674852]
   X-Ms-Error-Code: [BlobArchived]
   X-Ms-Request-Id: [c97536a8-e01e-00-50b7-c6f95f000000]
   X-Ms-Version: [2019-02-02]
```

**Comment by GoingOffRoading** [comment]
Similar Thread: https://forum.rclone.org/t/azure-storage-blob-overwrite-on-archive/9668/2

**Comment by ncw** [comment]
I've had a go at fixing this here.

It detects if we are updating an existing archive tier blob and deletes it first.

I find this change slightly concerning though - I wonder if it should be enabled with a flag?

Can you give it a go please?

[v1.54.0-beta.4943.d5d470526.fix-4819-azureblob-tier](https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4943.d5d470526.fix-4819-azureblob-tier/) on branch [fix-4819-azureblob-tier](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) (uploaded in 15-30 mins)

**Comment by GoingOffRoading** [comment]
> I've had a go at fixing this here.
> 
> It detects if we are updating an existing archive tier blob and deletes it first.
> 
> I find this change slightly concerning though - I wonder if it should be enabled with a flag?
> 
> Can you give it a go please?
> 
> [v1.54.0-beta.4943.d5d470526.fix-4819-azureblob-tier](https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4943.d5d470526.fix-4819-azureblob-tier/) on branch [fix-4819-azureblob-tier](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) (uploaded in 15-30 mins)


Given that there is potential consequence for deleting Archive-Tier data off of an Azure Blob, I would say that having this delete before option be a flag would be the right way to go.

I'm running somebody else's image of RClone in Docker (https://github.com/romancin/rclonebrowser-docker) so I will work on standing something up using the new branch when I get done with work... ETA 6 hrs

**Comment by GoingOffRoading** [comment]
@ncw  Looks like we have a winner!

What I executed:

- Downloaded the rclone-v1.54.0-beta.4943.d5d470526.fix-4819-azureblob-tier-windows-amd64.zip to my Windows machine
- Unzipped
- CMD'ed into the directory and created a placeholder rclone.conf
- Replaced rclone.conf with the one from my Linux host
- Did a dry run (Source, Remote, & File removed for privacy):

```
rclone.exe sync --delete-before --dry-run --progress SOURCE REMOTE
2020-11-30 13:28:37 NOTICE: FILE: Skipped copy as --dry-run is set
2020-11-30 13:28:37 NOTICE: FILE: Skipped copy as --dry-run is set
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Checks:              1128 / 1128, 100%
Transferred:            2 / 2, 100%
Elapsed time:         4.1s
```

- Did a full run (Source, Remote, & File removed for privacy):

```
rclone.exe sync --delete-before --progress SOURCE REMOTE
Transferred:        6.490G / 6.490 GBytes, 100%, 48.405 MBytes/s, ETA 0s
Checks:              1128 / 1128, 100%
Transferred:            2 / 2, 100%
Elapsed time:      2m20.0s
```

**Comment by GoingOffRoading** [comment]
Coming back to your "should this be a flag?"... Def YES

I only deleted a small handful of Archive-Tier files from my Azure Blob yesterday:

<img width="1015" alt="Screenshot 2020-11-30 134025" src="https://user-images.githubusercontent.com/4582826/100669028-a1fb6580-3311-11eb-9842-49833e7a380d.png">

If somebody had a LOT of files to delete, that could add up FAST.  While it's not RClone's responsibility to watch for data-store charges, it would be a very nice to have feature.

**Comment by ncw** [comment]
Thank you for testing.

What do you think the flag should be called, I suggest `--azure-blob-update-archive` maybe? This would be a config flag `update_archive`.

If `true` this flag would delete archive tier blobs before attempting to update them.

if `false` (the default) rclone should probably give an error `Can't update archive tier blobs without --azureblob-update-archive` - or maybe it could carry on anyway like it does at the moment and eventually return the `RESPONSE Stats: 409 This operation is not permitted on an archived blob` error.  Returning the `Can't update...` error is nice and user friendly so that is probably the thing to do provided we are certain you'll never be able to update an Archive tier blob.

Thoughts on the above?

**Comment by GoingOffRoading** [comment]
I wouldn't necessarily use the word 'update'.  'Update' involves updating the data which is not possible with a blob that is in Archive-Tier.

Maybe consider something like:

`--azure-blob-archive-tier-delete`   (that's long, I know)

**Comment by GoingOffRoading** [comment]
@ivandeex  sorry, I hit submit early... I've updated my comment if you wish to review

**Comment by ivandeex** [comment]
NP

**Comment by ncw** [comment]
I've had a go with the flag here - can you give it a test please @GoingOffRoading - comments on wording appreciated too :-)

[v1.54.0-beta.4957.bdee61e7e.fix-4819-azureblob-tier](https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4957.bdee61e7e.fix-4819-azureblob-tier/) on branch [fix-4819-azureblob-tier](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) (uploaded in 15-30 mins)

#### --azureblob-archive-tier-delete

Delete archive tier blobs before overwriting.

Archive tier blobs cannot be updated. So without this flag, if you
attempt to update an archive tier blob, then rclone will produce the
error:

    can't update archive tier blob without --azure-blob-archive-tier-delete

With this flag set then before rclone attempts to overwrite an archive
tier blob, it will delete the existing blob before uploading its
replacement.  This has the potential for data loss if the upload fails
(unlike updating a normal blob) and also may cost more since deleting
archive tier blobs early may be chargable.


- Config:      archive_tier_delete
- Env Var:     RCLONE_AZUREBLOB_ARCHIVE_TIER_DELETE
- Type:        string
- Default:     ""

**Comment by ncw** [comment]
Oops got the type of the flag wrong... Try this

[v1.54.0-beta.4957.4d4d91336.fix-4819-azureblob-tier](https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4957.4d4d91336.fix-4819-azureblob-tier/) on branch [fix-4819-azureblob-tier](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) (uploaded in 15-30 mins)

#### --azureblob-archive-tier-delete

Delete archive tier blobs before overwriting.

Archive tier blobs cannot be updated. So without this flag, if you
attempt to update an archive tier blob, then rclone will produce the
error:

    can't update archive tier blob without --azure-blob-archive-tier-delete

With this flag set then before rclone attempts to overwrite an archive
tier blob, it will delete the existing blob before uploading its
replacement.  This has the potential for data loss if the upload fails
(unlike updating a normal blob) and also may cost more since deleting
archive tier blobs early may be chargable.


- Config:      archive_tier_delete
- Env Var:     RCLONE_AZUREBLOB_ARCHIVE_TIER_DELETE
- Type:        bool
- Default:     false

**Comment by GoingOffRoading** [comment]
Sorry @ncw , it took me a little time to generate new files that created the Azure Blob Achieve Tier error...  I have those files now and am testing your branch

**Comment by GoingOffRoading** [comment]
Testing:  rclone-v1.54.0-beta.4957.4d4d91336.fix-4819-azureblob-tier-windows-amd64.zip

Excellent error message:  "2020-12-19 20:57:33 ERROR : FILENAME: Failed to copy: can't update archive tier blob without --azure-blob-archive-tier-delete"

Got an error:

```
C:\rclone2>rclone.exe sync --delete-before --progress --azure-blob-archive-tier-delete X:\SOURCE REMOTE:DIRECTORY
Error: unknown flag: --azure-blob-archive-tier-delete
Usage:
  rclone sync source:path dest:path [flags]

Flags:
      --create-empty-src-dirs   Create empty source dirs on destination after sync
  -h, --help                    help for sync

Use "rclone [command] --help" for more information about a command.
Use "rclone help flags" for to see the global flags.
Use "rclone help backends" for a list of supported services.

2020/12/19 20:58:39 Fatal error: unknown flag: --azure-blob-archive-tier-delete
```

**Comment by GoingOffRoading** [comment]
@ncw  let me know if you want me to test out any variants or etc

**Comment by x0b** [comment]
Shouldn't the flag be named `--azureblob-archive-tier-delete`?

**Comment by Isaystop** [comment]
I am reporting this the authorities you are violating my personal privacy

**Comment by Masamune3210** [comment]
> I am reporting this the authorities you are violating my personal privacy

What in the everlasting sun are you talking about?

**Comment by Isaystop** [comment]
Im talking about how everything that is being discussed is happening with my phones. All my drives and files are Magically disappearing and I started digging and this stupid site and all this bullshit is tied to the shot going on with my cell phones. I have never had an android but someone i have GitHub accounts linked to a android and google accounts. I have so many flow codes and hex codes sitting in my files it’s ridiculous. And currently I can’t fucking touch apps to open them.  I have already contacted the police and everyone’s users names have been reported. Now remove it from my clouds

**Comment by Masamune3210** [comment]
You are either a griefer or hopefully incredibly misunderstood. Nothing involving this project or this issue would have anything to do with your phone not working unless you ran something first, at which point it becomes user intention. 

Also, I sincerely hope you have hex in your files, as thats literally what a file is. If your files are messed up on your phone its likely either something wrong with your phone or its storage device. Maybe your sd card is failing or has become corrupted?

**Comment by Isaystop** [comment]
Nope. I’m neither. I am currently on my 6th Apple ID because all my files somehow keep geeting erased from my devices. My computer is gone. Because of these stupid codes. You all are writing these and people are using them to hurt others. I have lost every photos of my son and my grandmother all the texts between me and my ex. Gone. But magically nothing else was touched. And when I have tried to get everything back. He started to erase my
Social media accounts. I can’t change my Passwords. He see me changing them in real time. Like he’s watching my screen. It really really really is getting on my nerves. Please someone help me. I’ve delt with this for almost a year now. And I get a new phone. And it doesn’t stop. I’m so tired of loosing my digital footprint. And then i discovered Slipster is a user on her and a coder. And he Is dating my ex. And he is using multiple accounts that are linked to this one project. I’m not making this shit up. I swear to god.

**Comment by ivandeex** [comment]
@lsaystop

> all my files somehow keep geeting erased from my devices. My computer is gone. Because of these stupid codes.

Deeply sorry to hear that. Your story is very sad. If your issue is related to the rclone project, you can try asking on the forum or submit a new ticket. Yet this particular ticket has nothing to with that. Marking as offtopic. Sorry.

**Comment by GoingOffRoading** [comment]
@x0b yanking the dash from azure-blob is almost splitting hairs... Generally speaking, Azure will likely have more rclone compatible storage options in the future and it'd be good to keep those two entities separate (azure-blob) so that there's a clean schema in the future

Big picture -> doesn't really matter

**Comment by ncw** [comment]
@GoingOffRoading - the help message is wrong it should say --azureblob-archive-tier-delete - the name of the backend is `azureblob` so that is what the flags need to be called. I did prototype an `azurefile` backend but I didn't finish it!

I've corrected that - can you try this variant and if that works for you I'll merge it.

[v1.54.0-beta.4992.d01e711e2.fix-4819-azureblob-tier](https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4992.d01e711e2.fix-4819-azureblob-tier/) on branch [fix-4819-azureblob-tier](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) (uploaded in 15-30 mins)

**Comment by GoingOffRoading** [comment]
@ncw I'm not sure if the URL on   [https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4992.d01e711e2.fix-4819-azureblob-tier/]( https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4992.d01e711e2.fix-4819-azureblob-tier/) is broken but the page goes to "site can not be reached".

I also could not get to the releases page on the[ fix-4819](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) page.

I'm not sure if it's user-error on my behalf or if something else is going on.

I tried --azureblob-archive-tier-delete on the last release you linked me... The one a few comments up and I can confirm that it is working correctly/as expected.

**Comment by ncw** [comment]
> @ncw I'm not sure if the URL on https://beta.rclone.org/branch/fix-4819-azureblob-tier/v1.54.0-beta.4992.d01e711e2.fix-4819-azureblob-tier/ is broken but the page goes to "site can not be reached".
> 
> I also could not get to the releases page on the[ fix-4819](https://github.com/rclone/rclone/tree/fix-4819-azureblob-tier) page.
> 
> I'm not sure if it's user-error on my behalf or if something else is going on.

Sorry we had a bobble with the hosting. It should be accessible now.

> I tried --azureblob-archive-tier-delete on the last release you linked me... The one a few comments up and I can confirm that it is working correctly/as expected.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

Thanks for testing :-)
